# Import libraries

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import Data

In [15]:
data = pd.read_csv('data/heart.csv')

In [16]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [17]:
data.shape

(1025, 14)

# Preprocessing 
- Split our dataset into attributes and labels

In [21]:
X = data.iloc[:, :-1]
y = data.iloc[:, 13]

**The X contains all of the labels in our dataset, except for the target**  
**The y is our prediction variable, and it contains the last column**

# Train Test Split

We need to avoid overfitting.  In order to do this, we divide our data into training and test splits.  This will give us a better idea of how our algorithm performed during our testing phase. 
This way our algorithm is tested on un-seen data.  This is usual of production application. 

In [35]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

The above tells python to split the datasets into 80% training data and into 20% test data.  
This means that out of 1025 records, 820 will be used to train the data and 205 will be used to test it.  

# Feature Scaling

In [36]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

# Training and Prediction

In [37]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [38]:
y_pred = classifier.predict(X_test)

# Evaluation the Algorithm

In [39]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[81 19]
 [16 89]]
              precision    recall  f1-score   support

           0       0.84      0.81      0.82       100
           1       0.82      0.85      0.84       105

    accuracy                           0.83       205
   macro avg       0.83      0.83      0.83       205
weighted avg       0.83      0.83      0.83       205



**The algorithm was able to classify the 205 records for 0 with 84% accuracy, and for 1 with 82% accuracy.**